experimental/report/diagnostic.go

```diff
@@ -0,0 +1,235 @@
+// Copyright 2020-2024 Buf Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package report
+
+import (
+	"fmt"
+
+	"github.com/bufbuild/protocompile/experimental/internal"
+)
+
+// Level represents the severity of a diagnostic message.
+type Level int8
+
+const (
+	// Internal compiler error. Indicates a panic within the compiler.
+	ICE Level = 1 + iota
+	// Red. Indicates a semantic constraint violation.
+	Error
+	// Yellow. Indicates something that probably should not be ignored.
+	Warning
+	// Cyan. This is the diagnostics version of "info".
+	Remark
+
+	noteLevel // Used internally within the diagnostic renderer.
+)
+
+// Tag is a diagnostic tag: a machine-readable identification for a diagnostic.
+//
+// Tags should be lowercase identifiers separated by dashes, e.g. my-error-tag.
+// If a package generates diagnostics with tags, it should expose those tags as
+// constants.
+type Tag string
+
+// Apply implements [DiagnosticOption].
+func (t Tag) Apply(d *Diagnostic) {
+	if d.tag != "" {
+		panic("protocompile/report: set diagnostic tag more than once")
+	}
+
+	d.tag = t
+}
+
+// Diagnostic is a type of error that can be rendered as a rich diagnostic.
+//
+// Not all Diagnostics are "errors", even though Diagnostic does embed error;
+// some represent warnings, or perhaps debugging remarks.
+//
+// To construct a diagnostic, create one using a function like [Report.Error].
+// Then, call [Diagnostic.With] to apply options to it. You should at minimum
+// apply [Message] and either [InFile] or at least one [Snippet].
+type Diagnostic struct {
+	tag     Tag
+	message string
+
+	level Level
+
+	// sortOrder is used to force diagnostics to sort before or after each other
+	// in groups. See [Report.Sort].
+	sortOrder int
+	// The file this diagnostic occurs in, if it has no associated Annotations. This
+	// is used for errors like "file too big" that cannot be given a snippet.
+	inFile string
+
+	// A list of annotated source code spans in the diagnostic.
+	snippets           []snippet
+	notes, help, debug []string
+}
+
+// DiagnosticOption is an option that can be applied to a [Diagnostic].
+//
+// Nil values passed to [Diagnostic.With] are ignored.
+type DiagnosticOption interface {
+	Apply(*Diagnostic)
```

Comment:
Not wrong. I had contemplated merging DiagnosticOption and Diagnose... it's nice to not have two interfaces that basically do the same thing. At least, it would mean you could write `report.Error(report.Message(blah))`, but I would still want to keep `report.Errorf`.